[0-2 단계] 율무의 방탈출 예약 대기 미션 제출합니다.

kkiseug opened the pull request:
# 05/21 리뷰 반영!!
# [05/21 미션을 위한 Files changed 바로가기](https://github.com/woowacourse/spring-roomescape-waiting/pull/238/files/f388a7743331b9ad42dff62d90e3e674a7f82252..4caa072b729ca731a72260547574775d64ac79b9)

리뷰 반영하면서 많은 부분을 배운 것 같아요. 아쉬운 점은 시간이 촉박했다는 점입니다.. 그리고 테스트를 많이 작성하고 고민해보지 못해서 정말 아쉬워요. 현재는 테스트와의 격리성을 위해 각 테스트마다 데이터를 설정해주고 있어요. 그러다보니 오히려 테스트를 읽는데 어려운 것 같다는 생각도 들어요. 하지만, 픽스처를 사용하는 순간 생각의 범위가 파일 두 개 이상으로 늘어나니까 이것도 테스트를 읽기 어렵게 만드는 것 같더라고요. 어떤 기준점을 잡아야 할텐데 쉽지 않은 것 같습니다.

---
<br>

# [미션을 위한 Files changed 바로가기](https://github.com/woowacourse/spring-roomescape-waiting/pull/238/files/f388a7743331b9ad42dff62d90e3e674a7f82252..bbef528d02fa6e5007ddd0ca4bb08ccbed7d4814)

안녕하세요!! 샐리 🥗 샐러드가 생각나는 이름이라서 이모티콘을 붙여봤습니다.
리뷰 잘 부탁드려요!!!

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [내 프롤로그](https://prolog.techcourse.co.kr/studylogs/4377)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
```
[ ] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
[x] else 예약어를 쓰지 않는다.
[ ] 모든 원시 값과 문자열을 포장한다.
[ ] 일급 컬렉션을 쓴다.
[ ] 한 줄에 점을 하나만 찍는다.
[x] 줄여 쓰지 않는다(축약 금지).
[ ] 모든 엔티티를 작게 유지한다.
[ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
[ ] getter/setter/프로퍼티를 쓰지 않는다.
```
9/2 = 0.23점..

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
### 양방향 매핑의 기준
결론부터 말하면 양방향 매핑을 왜 쓰는지 모르겠어요. 양방향 매핑을 하는 순간 객체지향에서 더 멀어지는 기분이 들거든요. 객체지향에서 순환참조는 서로가 책임을 나눠 가진 것이 아닌가 생각하게 만듭니다. 근데 양방향 매핑 자체가 순환참조에 해당하는 것이죠. 여기부터 양방향 매핑이 맘에 안듭니다. 그리고, Member가 Reservation을 갖는 생각의 흐름이 좀 더 자연스러운데 DB는 이를 표현하기 어려우니 결국 Reservation이 Member를 가지게 되었다. 하지만. 이를 매핑한 JPA는 결국 DB의 패러다임을 객체로 표현해주기 위해서 양방향 매핑을 지원하지만.. 이러한 패러다임 불일치로 인해 객체지향과 오히려 멀어진 것이 아닌가하는 생각이 들어요.

그래서 양방향 매핑을 사용하려면 어떤 주인이 가진 책임을 수행하기 위한 매핑보다는 단순 주인을 조회하는 식에서 매핑이 좀 더 알맞을 수 있겠다. 라고 생각했습니다. 근데 재밌는건 이 관점을 Member에 적용시키기엔 너무 Reservation 도메인 책임이 되어 사실 양방향 매핑을 하지 않는게 올바른 상황이에요.

샐리의 기준을 들어보고 싶네요.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kkiseug committed:
feat: 프로젝트 구성

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
chore: Spring Data Jpa 의존성 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
chore: Jpa 관련 설정 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 기존 리포지토리 구현체 Spring Data jpa 구현체로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 테스트 스프링 설정 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 운영 초기 데이터 sql 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: ThemeService 테스트 config 주입으로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
fix: 테마 id 조회 메서드 네이밍 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 비즈니스 예외 처리 400 http 예외 코드로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 엔티티 멤버 속 VO record -> class

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 테스트 config 주입으로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 프론트 코드 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 레거시 코드 삭제

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 프론트 코드 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 예약 상태 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 예약 상태 추가로 인한 sql문 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 멤버와 예약 양방향 매핑

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 멤버별 예약 조회 기능 구현

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
test: 멤버별 예약 조회 테스트 작성

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
chore: 테스트 스키마 제거

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 멤버별 예약 조회 책임 Reservation 도메인으로 이동

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 예약 api 컨트롤러 역할별 분리

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
style: 컨벤션 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 멤버 예외 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
fix: 컨트롤러 예외 처리 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: VO를 record 불변 객체로 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
fix: Member 양방향 매핑 제거

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
fix: 멤버 양방향 의존성을 이용하던 코드 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug commented on src/main/java/roomescape/admin/domain/AdminRepository.java:
```diff
@@ -1,10 +1,12 @@
+import java.util.Optional;
 
-    Admin findById(Long id);
+public interface AdminRepository  {
```
본래 있던 AdminRepository가 도메인 패키지에 있기에 CrudRepository를 바로 참조하기 보다 이를 좀 더 확장성 있게 Jpa용 Repository를 만들고 구현체가 이를 합성해서 사용하도록 했어요. 복잡성이 오히려 증가하는 선택일까요?

dusdn1702 replied:
이렇게 했을 때 율무가 느끼는 장점은 무엇인가요?
도메인과 DB 단의 결합도를 낮추는 측면에서는 좋을 수 있지만 Repository를 도메인으로 보는 이유와 복잡성이 증가하는 것을 감내할 만한 장점이 무엇인지 궁금하네요 ㅎㅎ

kkiseug replied:
저는 리포지토리를 요구사항이 필요로 하는 도메인의 컬렉션을 만드는 관점에서 시작하였고, 이 과정에서 컬렉션이 필요로 하는 기능들을 추상화하여 인터페이스로 만들었기 때문에 이를 도메인에 가까운 계층이라고 생각했습니다. DDD의 영향이라 볼 수 있을 것 같습니다. 아직 DDD를 모르지만 이러한 철학이 마음에 들었습니다.

그리고, 도메인 리포지토리 인터페이스를 의존하고 있는 상황에서 당장 JPA의 Repository 인터페이스를 의존하게 되면 Repository가 너무 외부 의존성에 결합되는 것이 순수해보이지 않았습니다. 게다가 다른 구현체로 변경되는 유연성도 낮아진다고 생각했어요.

결과적으로 저는 이러한 구조로 유연성(DIP)도 가져갈 수 있었다고 생각하며, 도메인 계층의 인터페이스에서 필요한 메서드만 선언하기 때문에 ISP 원칙도 나름 지키고 있다고 생각합니다. 그리고, 가장 큰 이유는 사실 도메인 계층의 순수함을 지킬 수 있었다는 점입니다.

dusdn1702 replied:
그렇다면 JPA를 사용하는 이유는 무엇이라고 생각하시나요?
Jpa에서 `CrudRepository`로 추상화하여 기능을 제공하는 이유는 무엇일까요?
율무의 관점에서 JpaRepository가 도메인 계층의 순수함을 깨뜨린다고 생각한 이유가 궁금해요 ㅎㅎ
추가로 이렇게 구현했을 때의 단점은 복잡성 외에 어떤 부분이 있을까요?

kkiseug replied:
기존에는 저희가 미션에서 Jdbc API를 JdbcTemplate과 같은 것으로 직접 다루면서 ORM 기술을 미약하게나마 구현했다고 생각해요. 그래서 단순히 ORM 기술만을 위해서 JPA를 사용한다면 저는 굳이 사용할 필요 없다고 생각합니다. 상속같은 것이 나온다면 조금 달라지긴 하겠지만요... 하지만, JPA는 ORM 기술과 함께 영속성 컨텍스트, 연관관계 매핑 등의 기술도 지원하고 있기에 사용한다고 생각해요. 하지만, 이것도 결국 프레임워크이고 외부 의존성에 해당합니다. 더군다나 Jdbc를 사용할때조차 도메인에 집중한 인터페이스를 만들고, 해당 구현체를 분리했는데 왜 JPA를 사용한다고 해당 인터페이스에 직접적으로 침투하는걸 허용해야 할까?라고 느꼈고, 이를 허용하는 순간 도메인 계층의 순수함을 깨뜨린다고 생각했어요.

그럼 `CrudRepository`의 추상화 목적이 뭘까? 라는 질문이 남는데요. 저는 `EntityManager`의 사용을 좀 더 편리하게 하기위해서 만든 추상화 계층이라고 생각해요. 이 계층도 결국 JPA의 영속성 컨텍스트 없이는 동작하지 못하기 때문에 굉장히 강하게 결합되어 있어요. 그래서 더욱 설계의 방향을 지금처럼 잡은 것 같습니다.

게다가 요구사항에 따라 영속성 컨텍스트를 거치지 않고 차라리 직접적으로 예약을 입출력하는게 더 효율적인 상황이 생겼을 때, 기존에 다른 부분은 JPA를 유지하면서 상황에 따라 Jdbc 구현체를 주입해서 사용하는 방식을 생각할 수도 있을 거 같습니다. 하지만, 이정도의 상황이 정말 생길 가능성이 있을까에 대해서는 자신이 없긴 합니다.

dusdn1702 replied:
좋습니다! 저도 다른 인사이트를 많이 얻어가네요 👍 
지금은 요정도 생각을 가지고 jpa와 더 친해지고 복잡한 프로젝트를 만났을 때 어떠한 구조가 좋을지 또 고민해봐도 충분할 것 같아요.ㅎㅎ

kkiseug commented on src/main/java/roomescape/member/domain/Email.java:
```diff
@@ -1,19 +1,17 @@
 
-public class Email {
+@Embeddable
+public record Email(String email) {
```
VO를 record로 쓰는 것에 어떻게 생각하시나요? 저는 도메인 -> VO, 엔티티로 크게 나눌 수 있다고 생각합니다. 그래서 VO를 표현하는 방법 중 하나가 record일 뿐 해당 키워드를 쓰는순간 도메인 로직이 포함되면 안된다! 라는 관념이 적용되면 안된다고 생각합니다.

dusdn1702 replied:
저도 record를 사용할 수는 있다고 생각해요 ㅎㅎ
다만 율무가 보는 도메인 로직이란 어떤 것이 있나요? 다른 개발자들이 record 키워드를 사용했으면 이건 도메인 패키지 안에 있지만 vo라고 이해하길 기대하신다는 말씀이시죠?

kkiseug replied:
이메일 같은 경우는 도메인 로직이 포함되어 있지 않아요! 하지만, Name의 경우에는 이름의 길이를 제한하는데 이것도 도메인 로직으로 볼 수 있다고 생각합니다. 그래서 해당 주제를 고민했어요.

> 다른 개발자들이 record 키워드를 사용했으면 이건 도메인 패키지 안에 있지만 vo라고 이해하길 기대하신다는 말씀이시죠?

이 말이 맞습니다! 왜 이런 고민을 하냐면 vo에 대해서 도메인 로직이 포함되어야 한다 아니다로 갈리고, record에 대해서도 둘로 갈리더라고요! 그래서 뭔가 일반적인 시선이 정해지지 않아서 고민해보았습니다.

dusdn1702 replied:
우선 이 부분에 대해서 정해진 답은 없다고 생각해요. 💬
그렇다면 도메인의 vo를 어디까지 사용해도 되도록 허용할지, 이름 길이 제한과 같은 유효성은 어디서 진행할지에 대한 고민을 크루들과 해보고, 율무만의 기준을 세워나가면 좋을 것 같아요 👍

kkiseug replied:
저 같은 경우에는 VO도 도메인의 한 부분이라 생각해요. 왜냐면, 같은 객체라도 요구사항에 따라 VO가 될 수도 있고 엔티티가 될 수 있기 때문입니다. 예를 들어 편의점에서 물건을 구매하는 요구사항에서 돈은 VO로 동작합니다. 서로 다른 5000원을 구분없이 사용하더라도 해당 도메인 안에서는 제 역할을 할 수 있어요. 하지만, 조폐공사에서 돈을 관리 및 추적한다. 라는 요구사항에서는 서로 다른 5000원은 일련번호로 인해 정말 다른 객체 즉, 식별값을 가진 엔티티가 됩니다! 그래서 저는 VO도 도메인 객체의 일종으로 보기 떄문에 VO와 관련된 비즈니스 유효성 검사 같은 것을 VO에서 하는게 맞다고 생각합니다!

kkiseug commented on src/main/java/roomescape/admin/service/AdminService.java:
```diff
@@ -14,14 +15,16 @@ public AdminService(AdminRepository adminRepository) {
     public Admin findByEmail(final String email) {
-        return adminRepository.findByEmail(email);
+        return adminRepository.findByEmail(email)
+            .orElseThrow(() -> new BusinessException("관리자를 찾을 수 없습니다."));
```
orElseThrow()를 한번 더 추상화 해서 서비스 계층을 나누는 방법에 대해서 어떻게 생각하시나요?

dusdn1702 replied:
무슨 의미인지 이해하지 못했어요. 어디서 추상화되었다는 말씀이실까요??

kkiseug replied:
아! 지금은 추상화를 하진 않았어요. 제가 해결하고 싶은 문제점은 orElseThorw() 로직이 반복적으로 일어나는 점이었어요! 리포지토리에서 이 예외를 던져주는 방법도 생각할 수 있으나, 리포지토리는 해당 결과가 있든 없든 찾기만 한다면 책임의 의무를 다했다고 생각합니다.

그래서 서비스 단에서 계층을 나누어 분리할 순 없을까?를 여쭤보고 싶었어요!

dusdn1702 replied:
음 한번 나누어서 구현을 해보면 어때요? 그러고 그 구현의 장단점에 대해 같이 이야기해보면 좋겠어요.

kkiseug replied:
좋아용. 저두 해보는게 가장 좋다고 생각합니다.

kkiseug replied:
조회와 쓰기로 서비스를 나누고, 상위 계층으로 묶어서 사용하도록 해보았는데요!!! 이렇게 하니까 확실히 책임이 분리되는 것 같았습니다. 그리고 계층이 하나 더 생기고나니 상위 계층에서 부담없이 하위 계층을 의존할 수 있었던 것도 좋았어요. 또, 상위 계층은 결국 본인 도메인 컨트롤러에만 의존되니까 필요한 메서드만 정의하고 하위 계층에서 여러 도메인에 사용될 메서드를 구현하도록 해서 컨트롤러 -> 서비스 간의 복잡도가 많이 줄었다고 생각해요!

다만, 하나의 서비스가 3개로 나눠진만큼 관리 하기가 더 복잡해진건 있더라구요. 그래도 초반부터 이런 설계로 해나갔다면 어느정도 괜찮지 않았을까 싶긴해요! 테스트는 아직 제대로 작성하지 못해서 이를 개선하는 방법을 찾는 것도 필요할 것 같아요.

dusdn1702 replied:
오호 query와 command로 나누어주셨군요. 좋습니다.
지금은 가벼운 비즈니스 로직들을 다루고 있어서 복잡성만 늘어난 것처럼 느껴질 수 있지만 추후에 복잡한 프로젝트에서는 효과적으로 동작할 수 있을 거에요. 고생 많으셨습니다! 다음 미션에서 테스트도 같이 수정해보시죠 ㅎㅎ

kkiseug reviewed (commented):
# 궁금한 점

kkiseug commented:
@CodeRabbit 리뷰해

kkiseug commented:
@CodeRabbit 코드리뷰 해

kkiseug commented:
@CodeRabbit why some checks were not successful?

kkiseug commented:
@CodeRabbit 이제 다시 코드를 리뷰하고, 감상평을 한 줄로 요약해서 업로드 해봐

kkiseug commented:
@CodeRabbit Some checks were not successful을 좀 해결해봐

kkiseug commented:
@CodeRabbit /resume

kkiseug commented:
@CodeRabbit /resume

kkiseug commented:
@CodeRabbit post review comments plz.

dusdn1702 commented on src/main/java/roomescape/auth/exception/ForbiddenException.java:
```diff
@@ -0,0 +1,7 @@
+package roomescape.auth.exception;
```
auth를 common과는 나누고 admin, member, reservation 등의 도메인과는 같은 계층으로 두신 이유가 무엇인가요? AuthExceptionHandler 는 common에 존재하는게 의도하신 부분과 부합하나요??

kkiseug replied:
auth 또한 로그인, 회원가입 같은 하나의 도메인이라 생각해서 같은 계층으로 두었습니다. 그리고 AuthExceptionHandler는 exceptionHandler 자체가 컨트롤러 계층보다 상위에 존재하기도 하고 auth가 아닌 곳에서도 UnathorizedException을 발생시킬 가능성이 있다고 생각해서 두었어요. 근데, 사실 이런 일을 Interceptor와 ArgumentResolver에서 하고 있기에 또한, 이 둘은 이미 auth에 들어가 있어서 common에 있는 것이 어색해보이네요.

커밋: [6c77a91](https://github.com/woowacourse/spring-roomescape-waiting/pull/238/commits/6c77a916463d74cc121a897ab139eceeb3f4294b)

dusdn1702 replied:
음 패키지 분리에 대해서 더 고민해봐도 좋을 것 같아요.
도메인에 명확히 분리되지 않는 Exception이 추가되면 어느 패키지에 두실 건가요?
지금의 구조라면 MemberExceptionHandler도 따로 분리되어야하지 않을까요?
각 도메인별로 패키지를 관리할 때 어떤 단점이 있을 수 있을까요?
다른 방법으로는 어떤게 있을까요?

kkiseug replied:
AuthExceptionHandler를 지금 분리하는 것은 아직까지는 조금 보수적으로 갈 필요가 있다고 생각이 드네요.. 다만, 도메인에 명확히 분리되지 않는 Exception이 추가되는 것에 대해서는 어떤 상황이 생길지 잘 상상이 되진 않아요. 유추해본다면 시간과 날짜 중복에 대한 예외를 Reservation으로 볼 것인지 ReservationTime으로 볼 것인지에 대한 예시가 있을 것 같아요.

하지만, ReservationTime은 이미 Reservation과 너무 강결합 되어있다고 판단했습니다. 그래서 ReservationTime을 Reservation 패키지 안으로 옮겼습니다. 이렇게 도메인으로 관리하는 방법외에는 계층별로 나누고 이를 한 곳에 몰아넣는 방식이 있는데요. 지금 여기서 이 패키지 구조를 도입하기에는 너무 복잡해질 것 같다고 생각합니다.

dusdn1702 commented on src/main/java/roomescape/common/util/time/SystemDateTime.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class SystemDateTime implements DateTime {
```
인터페이스가 필요했던 이유가 무엇일까요??

kkiseug replied:
테스트와 유연성이 가장 큰 이유였어요. 특히 LocalDateTime.now()로 서비스 내에서 필드로 가지게 된다면, 테스트 작성 시에 날짜에 따라 실패하는 테스트가 작성된다고 생각했어요. 하지만, LocalDate.minusDays()와 같은 메서드로 충분히 항상 성공하는 테스트를 작성할 수 있더라고요.

유연성 같은 경우에는 현재 날짜로부터 일주일 간의 인기테마가 아니라 사용자가 날짜를 입력하면 그에 맞게 그 날짜로부터의 일주간의 인기테마를 보여줘야 하는 요구사항이 생길 수 있다고 생각했어요. 하지만, 이 질문에 답변하면서 생각하게 된 것이 입력 파라미터로 들어와도 서비스 안에서 해당 파라미터에 대해서 `SystemDateTime`을 사용하진 않을 것 같네요. 필요없는 코드였던 것 같습니다.

커밋: [2e3c5a3](https://github.com/woowacourse/spring-roomescape-waiting/pull/238/commits/2e3c5a3df1b1f5f9fe4c3eee585e9bc4ac10f3dd)

dusdn1702 commented on src/main/java/roomescape/member/domain/Email.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Objects;
+import roomescape.member.exception.EmailException;
+
+@Embeddable
```
이 어노테이션의 의미는 무엇인가요?
사용 시 주의할 점은 어떤게 있을까요?

kkiseug replied:
지금처럼 엔티티가 아닌 Value Object 같은 것을 테이블에 저장하고자 할 때, 객체를 그대로 저장할 순 없으며 이를 위한 테이블을 따로 만드는 것도 관리 대상이 늘어나기 때문에 좋지 않은 방법입니다. 그래서 `@Embeddable`과 `@Embedded` 애노테이션을 통해 엔티티 속 값 타입임을 명시할 수 있습니다!

주의할 점으로는 값 객체의 필드명이 그대로 테이블 컬럼으로 들어간다는 점이 있을 것 같아요! 직접 테스트를 해보니 많은 `@Embedded` 필드들이 있을 때 서로 중복되는 필드가 있다면 테이블이 만들어지지 않고 중복 예외가 발생하더라고요! 또한, insertable, updatable 설정에 따라 insert, update 쿼리가 나가지 않아 주의해야 할 것 같아요.

dusdn1702 replied:
좋습니다 ㅎㅎ 👍

dusdn1702 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -0,0 +1,56 @@
+public class Member {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
IDENTITY를 사용하신 이유가 궁금해요.

kkiseug replied:
H2는 기본적으로 SEQUENCE를 사용하는 것으로 알고 있고, IDENTITY를 사용하려면 url에 `;MODE=MySQL` 설정을 추가로 넣어줘야 하는 것으로 알고 있습니다. 하지만, JPA에서 엔티티 테이블을 만들때 `id BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY` 해당 옵션으로 id를 생성해 `MODE=MySQL` 설정없이 가능하도록 만들어줍니다. 그래서 IDENTITY를 사용해도 문제 없다고 생각했습니다.

오히려 SEQUENCE를 사용하게 되면 init.sql에서 초기 데이터를 생성하는 쿼리에서 예외가 발생하기도 하고, 추가적인 쿼리를 더 작성해주는 것이 오히려 비용적으로 손해라고 생각했습니다.

dusdn1702 commented on src/main/java/roomescape/member/exception/NameException.java:
```diff
@@ -0,0 +1,9 @@
+
+import roomescape.common.exception.BusinessException;
+
+public class NameException extends BusinessException {
```
Exception이 굉장히 많은데 이렇게 구현했을 때의 장단점은 무엇이 있을까요?

kkiseug replied:
저번 미션에서 막바지에 급하게 커스텀 예외를 대충 학습하고 적용했는데, 그땐 무슨 이유가 있었는데 다시 보니까 전혀 할 필요없는 커스텀 예외인 것 같네요.. 다시 한번 학습하고 적용해보겠습니다.

커밋: [9ea9166](https://github.com/woowacourse/spring-roomescape-waiting/pull/238/commits/9ea91665df4ce637759557133cea062f20415eec)

dusdn1702 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -0,0 +1,130 @@
+    private ReservationTime time;
+
+    @ManyToOne
+    private Theme theme;
```
ManyToOne으로 예약과 테마의 관계를 설정해주신 이유가 궁금해요.

kkiseug replied:
예약은 하나의 테마, 예약 시간, 멤버를 가질 수 있지만 반대 입장에서는 여러 개의 예약에 포함될 수 있습니다! 그래서 예약에 
 ManyToOne 관계로 설정했습니다.

dusdn1702 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -0,0 +1,130 @@
+    private Theme theme;
+
+    @ManyToOne
+    private Member member;
```
```
그래서 양방향 매핑을 사용하려면 어떤 주인이 가진 책임을 수행하기 위한 매핑보다는 단순 주인을 조회하는 식에서 매핑이 좀 더 알맞을 수 있겠다. 라고 생각했습니다. 근데 재밌는건 이 관점을 Member에 적용시키기엔 너무 Reservation 도메인 책임이 되어 사실 양방향 매핑을 하지 않는게 올바른 상황이에요.
```
이 부분에 대해 더 이야기 해보면 좋겠어요. Reservation의 책임이 무엇이어서 양방향 매핑을 적용하지 않으신건가요?
양방향 매핑으로 인해 객체지향에서 멀어지는 느낌이라는 부분에 어느정도 공감해요. 하지만 누릴 수 있는 장점도 분명히 있죠.
Reservation이 모든 연관관계를 갖고 있도록 구현하신 이유도 같이 설명해주시면 좋을 것 같네요 ㅎㅎ

kkiseug replied:
> Reservation의 책임이 무엇이어서 양방향 매핑을 적용하지 않으신건가요?
> 
> 이 관점을 Member에 적용시키기엔 너무 Reservation 도메인 책임이 되어 사실 양방향 매핑을 하지 않는게 올바른 상황이에요.

요구사항에서 내 예약 목록을 조회 할 수 있다고 했어요. 그래서 처음에는 요구사항에서 말하는 '내'가 '나의 예약'을 가지고 이를 조회할 수 있도록 처리하는게 이상적인 상황이라고 생각했습니다. 그렇기에 양방향 매핑을 이용했고, 이때부터 Member는 본인의 예약을 조회할 수 있게 됐습니다.

하지만, 이게 정말 좋은 설계일까? 객체지향적일까? 라고 생각했을 때 아니라고 생각했어요. 왜냐면 '내 예약 목록을 조회할 수 있다.'라는 요구사항이 원하는 것은 결국 예약인데, Member에 Reservation이 들어가면서 Member 도메인의 책임 되는 것 같은 기분이 들었습니다. 어쩌면 이런게 설계의 방향을 바꿀 수도 있다고 생각했어요.

오히려 이보다는 Reservation만 Member를 알고, Member에서는 Reservation을 모른채 내 예약 조회를 원한다면 오히려 ReservationService와 Reservation이 처리해주는게 더 자연스럽다 생각했어요! 그래서 양방향 매핑을 적용하지 않았습니다. 의도가 전달되었을까요?!

> 누릴 수 있는 장점도 분명히 있죠.

어떤 장점이 있을까요? 지금 미션 단계에서는 잘 모르겠어요.. ㅠ

> Reservation이 모든 연관관계를 갖고 있도록 구현하신 이유

Time, Theme의 관계는 요구사항에 의해서 고민없이 넣었던 것 같아요. 하지만, Reservation에 Member를 넣을 때는 Reservation이 너무 무거워 지는 것 같아 이를 ReservedMember와 같은 엔티티를 통해 해소하려고 했어요. 근데 이러면 제가 Admin, Member를 나눌 때 생각했던 역할이 아닌 '상태'를 기준으로 객체를 나눈 꼴이 되어버리더라고요. 그래서 Reservation에 연관관계를 설정했습니다.

dusdn1702 replied:
> 어떤 장점이 있을까요? 
처음 말씀 주신 것처럼 `'내'가 '나의 예약'을 가지고 이를 조회할 수 있도록 처리` 할 수 있죠. Reservation 과 Member의 관계를 명확히 표현할 수 있는 수단이 되기도 한다고 생각해요.
추후에 복잡한 쿼리들이 생길 때 호출의 횟수도 줄일 수 있을테구요.

왜 Reservation에 책임을 주었는지 충분히 이해했어요.ㅎㅎ 좋은 방향이라고 생각해요.
다만 Member 양방향도 적용해보셨듯, 전체적으로 미션에서 시도해볼 수 있는 부분들은 시도해봐도 좋다고 생각해요.
그 과정에서 얻을 수 있는 내용이 분명히 있을테니까요.

kkiseug replied:
> Reservation 과 Member의 관계를 명확히 표현할 수 있는 수단이 되기도 한다고 생각해요.

관계를 명확히 표현하기 위한 수단이라고 생각할 수도 있네요! 확실히 Member만 보면 Reservation과의 관계를 알기는 어렵긴 합니다. 좋은데요?!

> 전체적으로 미션에서 시도해볼 수 있는 부분들은 시도해봐도 좋다고 생각해요. 그 과정에서 얻을 수 있는 내용이 분명히 있을테니까요.

저도 해당 의견에 동의합니다! Admin, Member를 나눈 것도 이러한 생각에서 나눠보았어요 ㅎㅎ. 감사합니다~!

dusdn1702 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,123 @@
+        reservationRepository.findById(id)
+            .orElseThrow(() -> new BusinessException("멤버를 찾을 수 없습니다."));
+
+        reservationRepository.deleteById(id);
```
delete와 save가 동시에 들어올 땐 동작이 어떻게 되나요??

kkiseug replied:
delete와 동시에 save가 된다고 해도 서로 다른 데이터라면 영향이 없다고 생각했는데, 질문을 보고 다시 생각해보니 데이터 일관성에서 문제가 생길 수 있을 것 같아요. 미션을 하면서 트랜잭션을 고려하지 않았던 것 같습니다!

하지만, 궁금한 점은 이렇게 repository를 두 번 조회하는 경우에는 트랜잭션을 붙여주는 것이 맞지만 1회 조회, 쓰기에 대해서는 이미 Transactional이 내부적으로 처리되고 있더라구요. 그치만 결국 서비스 로직에서 여러 작업을 처리하면서 단 한번 repository 단에 접근하여 데이터를 조작 및 조회 하는데 그럼 서비스 로직 전체를 하나의 트랜잭션에 포함시켜야 하는지는 좀 고민됩니다.

dusdn1702 replied:
내부적으로 DB 단에서 처리되는 트랜잭션과 spring 어노테이션으로 관리되는 트랜잭션은 어떻게 다를까요?
이 부분 추가적으로 학습해보면 좋겠어요! 
단순히 repository 쿼리만 있다 하더라도 어노테이션을 통해 이 함수는 하나의 트랜잭션으로 관리된다고 명시할 수도 있겠죠.

kkiseug replied:
DB에서 처리되는 트랜잭션과 spring 애노테이션으로 관리되는 트랜잭션은 기본적으로 똑같이 DB에서의 트랜잭션입니다. 하지만, spring 애노테이션은 계층 범위의 개념이 더 들어갔더라구요. 그래서 DB 한정이 아닌 기능 동작에 ACID(그 중에서도 원자성)를 보장하는 애노테이션입니다!

그래서 제 코드에서 기능 동작의 원자성을 보장해야할 코드가 있는지 살펴보았어요. 특히, 이전에 적용했던 deleteById의 트랜잭션을 고민해보았는데요. 일단 제가 구현한 코드를 기준으로는 단건 삭제에 대해서 트랜잭션이 필요없다고 판단했어요! 왜냐면 findById에서의 트랜잭션이 조회의 원자성을 보장하고 있고, 이를 찾아서 영속성 컨텍스트에 들어간 후부터는 상태가 변화하든 상관없이 deleteById로 인해 삭제 상태가 됩니다. 그래서 트랜잭션이 없어도 각 메서드들의 트랜잭션으로 충분히 보호 가능하다고 생각했어요.

다만, CrudRepository의 구현체에서 deleteById를 했을 때 findById로 한번 찾아서 영속성 컨텍스트에 미리 저장을 해놓더라구요? 그래서 findById 메서드 호출을 제외하고 existsById로 존재하는지에 대해서만 판단하고 호출하도록 변경했습니다. 이외에는 단건 조회이기에 트랜잭션을 따로 적용해주지 않았습니다!

dusdn1702 replied:
오호 좋습니다!! 👍

dusdn1702 commented on src/main/java/roomescape/reservationTime/domain/ReservationTime.java:
```diff
@@ -0,0 +1,82 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private LocalTime startAt;
```
LocalTime으로 컬럼을 가져갈 때 DB에는 어떻게 저장이 되나요??

kkiseug replied:
이전에 Schema.sql이 있을 때는 VARCHAR로 저장했는데, JPA는 자동으로 time 타입으로 스키마를 생성하더라구요! 추가로 VARCHAR와 Time 중 어떤 것을 사용하는게 좋을지에 대한 주제로 논의가 있길래 확인해봤습니다! VARCHAR를 사용하면, 날짜에 대한 정렬이나 날짜 더하기, 빼기와 같은 것이 Time/Date 타입보다 어렵다고 하며 나중에 원하는 형식으로 바꾸기도 어려워진다고 합니다! 그리고, 이미 만들어진 기능이 있는데 안쓰면 개발자가 서운해 할 거라고 하더라고요. ㅋㅋㅋ

dusdn1702 replied:
ㅋㅋㅋㅋ좋네요.
`@Column` 선언은 전체적으로 없는데 안붙인 이유가 있나요?
붙이면 달라지는 부분은 무엇일까요?

kkiseug replied:
오, 있는지 몰랐네요.
사용하게 되면 제가 DDL을 커스텀 할 수 있네요!! nullable과 필드 사이즈를 제한해줬습니다!!

dusdn1702 commented on src/main/java/roomescape/theme/infrastructure/JpaThemeRepository.java:
```diff
@@ -0,0 +1,29 @@
+
+public interface JpaThemeRepository extends CrudRepository<Theme, Long> {
+
+    @Query(value = """
```
nativeQuery의 사용 이유가 궁금해요.
CrudRepository에 두신 이유는 무엇인가요?
이렇게 했을 때 어떤 문제가 발생할 수 있을까요?

kkiseug replied:
솔직하게 해당 부분은 페어를 하면서 사용 이유에 대해서 고민하진 않았던 것 같습니다. 이를 서비스 로직으로 풀어내려하니 쿼리를 사용하는 것보다 복잡했기에 해당 쿼리를 그대로 사용하는 방법으로 nativeQuery 있어 사용했습니다.

해당 질문을 보고 생각해보았는데요. CrudRepository 자체가 JPA의 장점을 활용하기 위한 인터페이스인데, nativeQuery를 사용함으로써 JPA의 데이터베이스 독립성이라는 장점을 잃어버리게 된 것 같습니다! 또한, 1차캐시를 먼저 찾는게 아니라 바로 DB에 쿼리문을 보내기때문에 영속성 컨텍스트의 이점도 못 얻을 것 같아요.

dusdn1702 replied:
좋습니다. 수정해주신 것 확인했어요.

dusdn1702 commented on src/test/java/roomescape/theme/service/ThemeServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+        );
+    }
+
+    static class ThemeConfig {
```
config 파일을 분리하지 않은 이유가 있나요??

kkiseug replied:
현재 제 테스트는 크게 도메인, 서비스, 통합 테스트로 나눌 수 있을 것 같아요. 여기서 Config 파일은 Service가 아니면 필요없는 클래스라고 생각해서 분리하지 않았습니다!

dusdn1702 replied:
오호 좋습니다!

kkiseug replied:
감사합니다~!

dusdn1702 commented on src/test/java/roomescape/theme/service/ThemeServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+class ThemeServiceTest {
+
+    @Autowired
+    private ThemeService themeService;
```
Service 테스트의 의미는 무엇인가요? 무엇을 검증하나요?
Repository 테스트와 연관 지어서 생각해보면 좋겠어요!

kkiseug replied:
저는 상황에 따라 다르지만, 지금처럼 JPA나 인기테마의 경우 쿼리 검증이 필요없다고 생각합니다! 그래서 Repository 테스트보다는 요구사항에 맞는 어떤 시나리오를 테스트 하는 것이 더 좋다고 생각해요. 그래서 Service 테스트가 이런 상황에서 더 의미있는 테스트가 된다고 생각하구요. 테스트를 하는 이유가 이러한 사실들을 검증하는 것이라 생각하기 때문입니다.

그래서 이 테스트에서 검증하는 것은 `어떤 조건에 대한 쿼리가 잘 동작하느냐?`가 아닌 `상위 10개의 테마를 잘 생성하고 있는가?`를 검증한다고 생각합니다. 좀 더 비즈니스적인 의미가 담긴 것..이라고 얘기할 수 있을 것 같아요. 제가 질문의 의도를 잘 이해한 것이 맞을까요?!

dusdn1702 replied:
넵 맞습니다
저는 `어떤 조건에 대한 쿼리가 잘 동작하느냐?` 가 repository로 검증되고,
그 검증된 쿼리 외에 로직이 ServiceTest에서 잘 수행되는지 검증하면 좋다고 생각해요.
다만 지금은 비즈니스 로직이 많지 않아 나누어지지 않은 것 같아요.
학습 차원에서 repository 테스트를 추가해봐도 좋겠어요.
여기서 검증된 쿼리는 service 테스트에서 어떻게 사용할지도 같이요!

kkiseug replied:
> 어떤 조건에 대한 쿼리가 잘 동작하느냐? 가 repository로 검증

저도 해당 의견에는 동의합니다! 그래서 커스텀 쿼리(인기 테마를 가져오는 것)에 대해서만 테스트를 진행해보았어요. 왜냐면 JPA를 사용하게 되었으니 기본적인 부분들은 JPA가 잘 구현해줬을거라 믿기 때문입니다! 다만, 인기 테마를 가져오는 부분이 Repository와 Service가 사실상 비슷하게 동작하기 때문에 어떤 점에서 해당 테스트를 나누는게 의미있는지는 아직 체감하지 못한 것 같습니다.

샐리의 말대로
> 지금은 비즈니스 로직이 많지 않아 나누어지지 않은 것 같아요.

이게 맞는 것 같아요. 오히려 지금 같은 상황에서는 Service만으로도 충분하다고 생각했습니다. 다만, 학습 차원이라고 한다면 동의합니다!

dusdn1702 reviewed (changes requested):
안녕하세요 율무!
jpa 미션 잘 수행해주셨네요. 고생 많으셨습니다.
남겨둔 코멘트 확인 부탁드릴게요~

dusdn1702 commented:
아 그리고 rabbit ai를 굉장히 잘 쓰시네요 ㅋㅋㅋㅎㅎ

kkiseug committed:
feat: 서비스 계층 트랜잭션 애노테이션 추가

kkiseug committed:
refactor: NativeQuery -> JPQL로 변경

kkiseug committed:
refactor: AuthExceptionHandler 패키지 이동

kkiseug committed:
refactor: DateTime 인터페이스 및 구현체 제거

- 테스트 코드 수정
- data.sql 수정

kkiseug committed:
refactor: 커스텀 예외 제거

kkiseug reviewed (commented):
의도를 많이 물어봐주셔서 고민할 시간을 갖게되어 좋았어요!!!!! 감사합니다.
커스텀 예외는 진짜 다시 한번 학습해보고 적용해야겠어요.
테스트 코드도 좀 더 알아봐야 할 것 같습니다.

kkiseug commented:
@CodeRabbit 코드리뷰 해

kkiseug commented:
@CodeRabbit 내 전체 코드에 한줄평 남겨줘

kkiseug commented:
@CodeRabbit 현재 요구사항에서 Admin과 Member의 분리가 적절하다고 생각해? 본인의 생각과 적절하다와 적절하지 않다의 입장에서 장점, 단점을 정리해줘

kkiseug commented:
@CodeRabbit 테스트 코드 위주로 리뷰하면서 data.sql에 대한 내용과 테스트 픽스처 적용에 관한 너의 생각도 들려줘

kkiseug commented on src/main/java/roomescape/auth/exception/AuthExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import roomescape.common.exception.handler.dto.ExceptionResponse;
+
+@RestControllerAdvice
+public class AuthExceptionHandler {
+
+    @ExceptionHandler(UnauthorizedException.class)
+    public ResponseEntity<ExceptionResponse> handleUnauthorize(final UnauthorizedException exception, final HttpServletRequest request) {
+        ExceptionResponse exceptionResponse = new ExceptionResponse(
+                HttpStatus.UNAUTHORIZED.value(), "[ERROR] " + exception.getMessage(), request.getRequestURI()
+        );
+
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED.value()).body(exceptionResponse);
+    }
+
+    @ExceptionHandler(ForbiddenException.class)
+    public ResponseEntity<ExceptionResponse> handleForbidden(final ForbiddenException exception, final HttpServletRequest request) {
+        ExceptionResponse exceptionResponse = new ExceptionResponse(
+                HttpStatus.FORBIDDEN.value(), "[ERROR] " + exception.getMessage(), request.getRequestURI()
+        );
+
+        return ResponseEntity.status(HttpStatus.FORBIDDEN.value()).body(exceptionResponse);
+    }
+}
```
좋은 아이디어지만 너무 많은 힌트를 주지마

kkiseug replied:
@CodeRabbit 대답

dusdn1702 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -44,62 +50,74 @@ public ReservationService(
+        Theme theme = themeRepository.findById(request.themeId())
+            .orElseThrow(() -> new ThemeException("테마를 찾을 수 없습니다."));
+        Member member = memberRepository.findById(memberId)
+            .orElseThrow(() -> new BusinessException("멤버를 찾을 수 없습니다."));
```
member만 왜 BusinessException인가요??

kkiseug replied:
커스텀 예외 학습을 좀 더 고민하고 반영하려고 일단 `BusinesssException`을 제외하고 모든 커스텀 예외를 지웠어요!

dusdn1702 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -44,62 +50,74 @@ public ReservationService(
+        validateExistDuplicateReservation(reservations, time);
 
-        Reservation reservation = Reservation.createWithoutId(request.date(), time, theme, member);
+        Reservation reservation = Reservation.createWithoutId(request.date(), time, theme, member, Status.RESERVED);
```
Status는 왜 필요한가요?? enum 이름을 명확히 해주면 좋겠어요.

kkiseug replied:
Status는 예약 상태를 나타내기 위한 Enum입니다! 현재로써는 크게 기능하는 코드는 아닙니다. 다만, 2단계에서 요구사항이 필요로 했기에 나온 Enum 입니다! Enum 이름을 좀 더 명확히 하기위해 ReservationStatus 네이밍으로 변경했어요.

처음엔 Reservation 내부에서 Reservation.ReservationStatus라는 방식이 좀 어색해보여서 그대로 유지하는 방향으로 가려고 했어요. 근데, 제가 ReservationTime을 만들때도 똑같은 고민을 한 적이 있습니다. 그때도 ReservationTime 대신 Time과 같은게 좀 더 일반적이고 Reservation 내부에 들어가서 의미를 나타내는 것이 아닌가라고 생각했었습니다. 근데, 그렇게 하면 '예약 시간'이라는 의미가 퇴색되더라고요. 해당 시간이 예약 시간이라는 도메인적 유의성을 가지는 것이 아니라 오히려 common이나 일반적인 시간의 의미를 가지게 되었다?라고 말할 수 있을 것 같아요. 그래서 Status도 동일하게 도메인적 유의성을 띄게 만들어야 했다고 생각합니다! 감사합니다.

dusdn1702 replied:
아하 좋습니다! 👍

dusdn1702 reviewed (changes requested):
안녕하세요 율무 
피드백 잘 반영해주시고, 많은 고민 해주신 부분들 확인하고 답변 남겨두었습니다. 👍 
한번 더 고민해봤으면 하는 내용들 있어 확인 부탁드릴게요!

kkiseug committed:
refactor: Repository 삭제 로직 변경

- Theme, Reservation, ReservationTime의 존재 여부 확인 로직 추가
- 불필요한 트랜잭션 어노테이션 제거

kkiseug committed:
refactor: 엔티티 필드에 JPA @Column 제약 조건 추가

- nullable=false로 필수값 설정 및 길이 제한 추가 (Password, Name, Theme)

kkiseug committed:
refactor: 인기 테마 쿼리 수정

- limit 제한으로 DB단에서 많은 데이터를 읽지 않도록 리팩터링

kkiseug committed:
refactor: 불필요한 static 팩토리 메서드 삭제

- 생성자를 통해 생성하도록 변경

kkiseug committed:
test: 기간별 인기테마 테스트 추가

- data.sql 삭제

kkiseug committed:
refactor: Status를 ReservationStatus로 리팩토링

- 'Status' 클래스 및 관련변수를 'ReservationStatus'로 이름 변경

kkiseug committed:
refactor: 서비스 계층을 Command와 Query로 분리하여 책임 분리

- 기존 단일 서비스 클래스를 CommandService와 QueryService로 분리
- FacadeService를 추가해 CommandService와 QueryService를 통합
- 컨트롤러 및 관련 클래스에서 새로운 서비스 계층 호출로 변경

kkiseug committed:
test: 서비스 계층 테스트 Facade 계층 테스트로 변경

- 테스트 클래스와 메서드 내 호출 서비스 이름 변경

kkiseug committed:
refactor: 패키지 구조 변경 및 AuthExceptionHandler 삭제

- AuthExceptionHandler 삭제로 불필요한 코드 제거
- `reservationTime` 및 기타 관련 패키지의 naming을 `reservation.time`으로 수정
- 서비스 계층 패키지를 `application`으로 이동하여 구조 개선

kkiseug committed:
feat: 예외 처리 추가 및 코드 리팩토링

- UnauthorizedException, ForbiddenException에 대한 예외 처리 추가
- 공통 예외 응답 생성을 위한 `createErrorResponse` 메서드 추가
- 코드 가독성을 위한 형식 및 구조 개선

kkiseug reviewed (commented):
리뷰 감사합니다~!

dusdn1702 commented on src/main/java/roomescape/member/application/MemberFacadeService.java:
```diff
@@ -0,0 +1,43 @@
+import roomescape.member.presentation.dto.MemberResponse;
+
+@Service
+public class MemberFacadeService {
```
나중에 생각해보면 될 것 같지만..ㅎㅎ
Facade라는 이름은 어떤 의미일까요?

dusdn1702 reviewed (approved):
안녕하세요 율무!
피드백을 나누며 많은 고민을 하고 저도 많이 배운 것 같아요. 고생 많으셨습니다.
이번 단계에서 학습할 내용은 충분히 하신 것 같아 머지하도록 할게요.
다음 단계에서 뵙겠습니다 👍

dusdn1702 commented:
pr에 리뷰 file changes 첨부해주신 것 너무 편했어요. 감사합니다!
